### PR TITLE
PIMS-134 Adding Geocoder endpoint

### PR DESCRIPTION
**Pull request by Fosol**
## Description

We now have access to Geocoder and can search for valid addresses based on entering a partial address.

### Endpoints
- `/api/tools/geocoder/addresses`

### CODE REVIEW

File: backend/api/Areas/Keycloak/Mapping/Role/RoleMap.cs [hunk 1]
```diff
@@ -10,7 +11,7 @@ public class RoleMap : IRegister
         public void Register(TypeAdapterConfig config)
         {
             config.NewConfig<Entity.Role, Model.RoleModel>()
-                .Map(dest => dest.Id, src => src.Id)
+                .Map(dest => dest.KeycloakGroupId, src => src.KeycloakGroupId)
```
Comment (Fosol) [inline backend/api/Areas/Keycloak/Mapping/Role/RoleMap.cs:14]: Fixing a bug with admin roles.

File: backend/api/Areas/Keycloak/Mapping/User/UserMap.cs [hunk 2]
```diff
@@ -22,7 +23,7 @@ public void Register(TypeAdapterConfig config)
                 .Map(dest => dest.Email, src => src.Email)
                 .Map(dest => dest.Note, src => src.Note)
                 .Map(dest => dest.Agencies, src => src.Agencies)
-                .Map(dest => dest.Roles, src => src.Roles)
+                .Map(dest => dest.Roles, src => src.Roles.Select(r => r.Role))
```
Comment (Fosol) [inline backend/api/Areas/Keycloak/Mapping/User/UserMap.cs:26]: Fixing bug with admin roles.

File: backend/api/Startup.cs [hunk 3]
```diff
@@ -93,17 +92,32 @@ public void ConfigureServices(IServiceCollection services)
                     member.Type.IsGenericType &&
                     member.Type.GetGenericTypeDefinition() == typeof(ICollection<>));
             });
+            services.Configure<JsonSerializerOptions>(this.Configuration.GetSection("Serialization:Json"));
```
Comment (Fosol) [inline backend/api/Startup.cs:95]: Adding more configuration options to support serialization for making HTTP requests.

File: backend/api/Startup.cs [hunk 4]
```diff
@@ -170,29 +184,18 @@ public void ConfigureServices(IServiceCollection services)
...
+            services.AddScoped<IOpenIdConnectRequestClient, OpenIdConnectRequestClient>();
 
             services.AddHealthChecks()
                 .AddCheck("liveliness", () => HealthCheckResult.Healthy())
                 .AddSqlServer(builder.ConnectionString, tags: new[] { "services" });
 
-            //TODO: Add a health check for keycloak connectivity.
-            // services.AddHealthChecksUI();
-
-            services.AddMvcCore()
```
Comment (Fosol) [inline backend/api/Startup.cs:186]: Moved this up to keep things organized.

File: backend/api/geocoder.json [hunk 5]
```diff
@@ -0,0 +1,3 @@
+{
+  "Host": "https://geocoder.api.gov.bc.ca"
```
Comment (Fosol) [inline backend/api/geocoder.json:2]: This URL requires an `apikey` to access.  The other URLs are public.

File: backend/core/Exceptions/HttpClientRequestException.cs [hunk 6]
```diff
@@ -2,12 +2,12 @@
...
 
-namespace Pims.Keycloak.Exceptions
+namespace Pims.Core.Exceptions
 {
     /// <summary>
-    /// KeycloakRequestException class, provides a way to express HTTP request exceptions that occur.
+    /// HttpClientRequestException class, provides a way to express HTTP request exceptions that occur.
     /// </summary>
-    public class KeycloakRequestException : HttpRequestException
+    public class HttpClientRequestException : HttpRequestException
```
Comment (Fosol) [inline backend/core/Exceptions/HttpClientRequestException.cs:10]: Generic exception for HTTP requests.

File: backend/core/Http/HttpRequestClient.cs [hunk 7]
```diff
@@ -0,0 +1,317 @@
+using Microsoft.Extensions.Options;
+using Pims.Core.Exceptions;
+using System;
+using System.Net.Http;
+using System.Text.Json;
+using System.Threading.Tasks;
+
+namespace Pims.Core.Http
+{
+    public class HttpRequestClient : IHttpRequestClient
```
Comment (Fosol) [inline backend/core/Http/HttpRequestClient.cs:10]: Generic implementation that we can reuse.

File: backend/geocoder/GeocoderService.cs [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        /// Sends an HTTP request to Geocoder for addresses that match the specified 'parameters'.
+        /// </summary>
+        /// <param name="parameters"></param>
+        /// <param name="outputFormat"></param>
+        /// <returns></returns>
+        public async Task<FeatureCollectionModel> GetSiteAddressesAsync(AddressesParameters parameters, string outputFormat = "json")
+        {
+            var uri = new Uri($"{GenerateUrl(this.Options.Sites.AddressesUrl, outputFormat)}");
+            var url = QueryHelpers.AddQueryString(uri.AbsoluteUri, parameters.ToQueryStringDictionary());
+            return await this.Client.GetAsync<FeatureCollectionModel>(url);
```
Comment (Fosol) [inline backend/geocoder/GeocoderService.cs:79]: Only currently implementing one of the possible Geocoder endpoints at this point in time.

File: backend/api/Areas/Tools/Mapping/Geocoder/AddressMap.cs [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+                address.Append($" {properties.StreetType}");
+            return address.ToString();
+        }
+
+        private string GetCity(GModel.PropertyModel properties)
+        {
+            return properties.LocalityType == "City" ? properties.LocalityName : null;
+        }
+
+        private double GetLatitude(GModel.GeometeryModel goemetry)
```
Comment (asanchezr) [inline backend/api/Areas/Tools/Mapping/Geocoder/AddressMap.cs:55]: /nit: `goemetry` -> `geometry`

Comment (Fosol) [reply backend/api/Areas/Tools/Mapping/Geocoder/AddressMap.cs]: fixed

File: backend/geocoder/Models/GeometeryModel.cs [hunk 10]
```diff
@@ -0,0 +1,13 @@
+using System.Collections.Generic;
+
+namespace Pims.Geocoder.Models
+{
+    public class GeometeryModel
```
Comment (asanchezr) [inline backend/geocoder/Models/GeometeryModel.cs:5]: `GeometeryModel` -> `GeometryModel`

Comment (Fosol) [reply backend/geocoder/Models/GeometeryModel.cs]: fixed

**Pull request merged**